File: gui-tools/src/gpstudio_lib/camera/property.cpp
/****************************************************************************
** Copyright (C) 2016 Dream IP
** 
** This file is part of GPStudio.
**
** GPStudio is a free software: you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation, either version 3 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program.  If not, see <http://www.gnu.org/licenses/>.
**
****************************************************************************/

#include "property.h"

#include <QDebug>

#include "model/model_property.h"
#include "model/model_flow.h"

#include "camera.h"

Property::Property(QString name)
    : _name(name)
{
    _parent = NULL;
    _type = Group;
    _row = 0;
    _bits = 0;
    _modelProperty = NULL;
    _fixed = false;
    _const = false;
}

Property::~Property()
{
    foreach (Property *property, _subProperties)
        delete property;
    foreach (PropertyEnum *propertyEnum, _enumsMap)
        delete propertyEnum;
}

const QString &Property::name() const
{
    return _name;
}

void Property::setName(const QString &name)
{
    _name = name;
}

const QString &Property::caption() const
{
    return _caption;
}

void Property::setCaption(const QString &caption)
{
    _caption = caption;
}

const QVariant &Property::value() const
{
    return _value;
}

void Property::setValue(bool value)
{
    setValue(QVariant(value));
}

void Property::setValue(int value)
{
    setValue(QVariant(value));
}

void Property::setValue(const QVariant &value)
{
    bool valueChangedb = false;
    if(value != _value && _value.isValid())
        valueChangedb=true;
    _value=value;

    if(_type==Matrix)
    {
        if(value.type()==QVariant::List)
        {
            int x=0;
            foreach (QVariant line, value.toList())
            {
                if(line.type()==QVariant::List)
                {
                    int y=0;
                    foreach (QVariant item, line.toList())
                    {
                        QString key = QString("m%1%2").arg(x).arg(y);
                        if(_subPropertiesMap.contains(key)) _subPropertiesMap[key]->setValue(item.toInt());
                        y++;
                    }
                }
                x++;
            }
        }
    }
    else
    {
        if(!_enumsMap.isEmpty())
        {
            if(_enumsMap.contains(value.toString()))
            {
                setBits(_enumsMap[value.toString()]->value().toUInt());
            }
        }
        else
        {
            setBits(value.toUInt());
        }
    }

    if(valueChangedb)
        emit valueChanged(QVariant(value));

    ScriptEngine::getEngine().evalPropertyMap(_onchange);
}

uint Property::bits() const
{
    return _bits;
}

void Property::setBits(const uint bits)
{
    if(bits != _bits)
    {
        _bits = bits;
        emit bitsChanged(bits);
    }
    else _bits = bits;
}

void Property::eval()
{
    if(_propertyMap.isEmpty())
        return;

    QScriptValue value = ScriptEngine::getEngine().eval(_propertyMap);
    if(!value.isError())
        setValue(value.toVariant());
}

const QVariant &Property::min() const
{
    return _min;
}

void Property::setMin(const QVariant &min)
{
    _min = min;
}

const QVariant &Property::max() const
{
    return _max;
}

void Property::setMax(const QVariant &max)
{
    _max = max;
}

const QVariant &Property::step() const
{
    return _step;
}

void Property::setStep(const QVariant &step)
{
    _step = step;
}

const QMap<QString, PropertyEnum *> &Property::enumsMap() const
{
    return _enumsMap;
}

const QList<PropertyEnum *> Property::enums() const
{
    return _enumsMap.values();
}

Property::Type Property::type() const
{
    return _type;
}

QVariant::Type Property::variantType() const
{
    switch (_type)
    {
    case Property::String:
    case Property::StringType:
        return QVariant::String;
    case Property::Int:
    case Property::SInt:
        return QVariant::Int;
    case Property::Bool:
        return QVariant::Bool;
    case Property::Enum:
        return QVariant::StringList;
    case Property::Matrix:
        return QVariant::StringList;
    default:
        return QVariant::Invalid;
    }
}

void Property::setType(const Property::Type &type)
{
    _type = type;
}

const QString &Property::propertymap() const
{
    return _propertyMap;
}

void Property::setPropertymap(const QString &propertymap)
{
    _propertyMap = propertymap;
}

const QString &Property::onchange() const
{
    return _onchange;
}

void Property::setOnchange(const QString &onchange)
{
    _onchange = onchange;
}

Property &Property::operator[](const QString &name)
{
    return *path(name);
}

const QList<Property *> &Property::subProperties() const
{
    return _subProperties;
}

Property *Property::parent() const
{
    return _parent;
}

const ModelProperty *Property::modelProperty() const
{
    return _modelProperty;
}

QString Property::blockName() const
{
    if(_type==BlockType)
        return _name;

    if(!_parent)
        return QString();

    return _parent->blockName();
}

QString Property::getPath() const
{
    if(_type==BlockType)
        return QString();

    if(!_parent)
        return QString();

    if(_parent->type()!=BlockType)
        return _parent->getPath()+"."+_name;
    return _name;
}

void Property::setParent(Property *parent)
{
    _parent = parent;
}

bool Property::isFixed() const
{
    return _fixed;
}

bool Property::isConst() const
{
    return _const;
}

Property::Mode Property::mode() const
{
    return _mode;
}

void Property::setMode(Property::Mode mode)
{
    _mode = mode;
    foreach (Property *property, _subProperties)
        property->setMode(mode);
}

int Property::row() const
{
    return _row;
}

void Property::setRow(int row)
{
    _row = row;
}

Property *Property::path(const QString &path) const
{
    if(path.isEmpty() || path=="value" || path=="bits")
        return (Property *)this;
    int index = path.indexOf(".");
    if(index==-1)
    {
        if(_subPropertiesMap.contains(path))
            return _subPropertiesMap[path];
        else return NULL;
    }
    if(_subPropertiesMap.contains(path.left(index)))
        return _subPropertiesMap[path.left(index)]->path(path.mid(index+1));
    return NULL;
}

QVariant Property::property(const QString &path) const
{
    Property *prop = this->path(path);
    if(prop)
        return prop->value();
    return QVariant();
}

QStringList Property::dependsProperties() const
{
    return ScriptEngine::dependsProperties(_propertyMap);
}

const QMap<QString, Property *> &Property::subPropertiesMap() const
{
    return _subPropertiesMap;
}

void Property::addSubProperty(Property *property)
{
    property->setParent(this);
    property->setRow(_subProperties.count());
    _subPropertiesMap.insert(property->name(), property);
    _subProperties.append(property);
    emit subPropertyChange();
}

void Property::removeSubProperty(Property *property)
{
    _subPropertiesMap.remove(property->name());
    _subProperties.removeOne(property);
    delete property;
    emit subPropertyChange();
}

void Property::removeAllSubProperties()
{
    foreach (Property *property, _subProperties)
        delete property;
    _subPropertiesMap.clear();
    _subProperties.clear();
}

Property *Property::fromModelProperty(const ModelProperty *modelProperty)
{
    Property *propertyParam = new Property(modelProperty->name());
    propertyParam->setCaption(modelProperty->caption());
    propertyParam->setOnchange(modelProperty->onchange());
    propertyParam->setPropertymap(modelProperty->propertymap());
    propertyParam->_modelProperty = modelProperty;
    if(!modelProperty->propertyEnums().empty())
    {
        foreach (ModelPropertyEnum *blockPropertyEnum, modelProperty->propertyEnums())
        {
            PropertyEnum *propertyEnum = new PropertyEnum(blockPropertyEnum->name(), blockPropertyEnum->caption(), blockPropertyEnum->value());
            propertyParam->_enumsMap.insert(blockPropertyEnum->name(), propertyEnum);
        }
        propertyParam->setType(Property::Enum);
        propertyParam->setValue(modelProperty->value());
    }
    if(modelProperty->type()=="int" || modelProperty->type()=="sint")
    {
        if(modelProperty->type()=="int") propertyParam->setType(Int);
        if(modelProperty->type()=="sint") propertyParam->setType(SInt);
        propertyParam->setValue(modelProperty->value().toInt());
        propertyParam->setMin(modelProperty->min());
        propertyParam->setMax(modelProperty->max());
        propertyParam->setStep(modelProperty->step().toInt());
    }
    if(modelProperty->type()=="matrix")
    {
        propertyParam->setType(Matrix);
        propertyParam->setValue(QVariant(modelProperty->value()).toInt());
    }
    if(modelProperty->type()=="bool")
    {
        propertyParam->setType(Bool);
        propertyParam->setValue(QVariant(modelProperty->value()).toBool());
    }
    if(modelProperty->type()=="group") propertyParam->setType(Group);
    if(modelProperty->type()=="string")
    {
        propertyParam->setType(String);
        propertyParam->setValue(modelProperty->value());
    }
    if(modelProperty->type()=="flowtype")
    {
        propertyParam->setType(FlowDataType);
        propertyParam->setValue(modelProperty->value());
    }

    // sub properties
    foreach (ModelProperty *subBlockProperty, modelProperty->properties())
    {
        propertyParam->addSubProperty(Property::fromModelProperty(subBlockProperty));
    }

    return propertyParam;
}

Property *Property::fromModelFlow(const ModelFlow *modelFlow)
{
    Property *propertyFlow = new Property(modelFlow->name());
    propertyFlow->setCaption(modelFlow->name());
    propertyFlow->setType(FlowType);
    propertyFlow->setValue((int)modelFlow->size());
    propertyFlow->_fixed = true;
    propertyFlow->_const = true;

    // sub properties
    foreach (ModelProperty *subBlockProperty, modelFlow->properties())
    {
        propertyFlow->addSubProperty(Property::fromModelProperty(subBlockProperty));
    }

    return propertyFlow;
}

Property *Property::fromModelBlock(const ModelBlock *modelBlock)
{
    Property *propertyBlock = new Property(modelBlock->name());
    propertyBlock->setCaption(modelBlock->name() + " (" + modelBlock->driver() + ")");
    propertyBlock->setType(BlockType);

    return propertyBlock;
}

Property *Property::fromModelParam(const ModelParam *modelParam)
{
    Property *propertyParam = new Property(modelParam->name());
    propertyParam->setCaption(modelParam->name());
    propertyParam->setType(StringType);
    propertyParam->setValue(modelParam->value().toString());
    propertyParam->_fixed = true;
    propertyParam->_const = true;

    return propertyParam;
}

Property *Property::fromModelClock(const ModelClock *modelClock)
{
    Property *propertyParam = new Property(modelClock->name());
    propertyParam->setCaption(modelClock->name());
    propertyParam->setType(Clock);
    propertyParam->setValue(modelClock->typical());
    propertyParam->_fixed = true;
    propertyParam->_const = true;

    return propertyParam;
}

File: scripts/model/property.php
<?php
/*
 * Copyright (C) 2016 Dream IP
 * 
 * This file is part of GPStudio.
 *
 * GPStudio is a free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

require_once("propertyenum.php");

/**
 * Property is contained into the Block::$properties and Flow::$properties
 * @brief The Property class is used to define high level properties
 * @see Block Param ParamBitfield Flow
 * @ingroup base
 */
class Property
{
    /**
     * @brief Name of the property
     * @var string $name
     */
    public $name;

    /**
     * @brief Caption of the property for high level interface
     * @var string $caption
     */
    public $caption;

    /**
     * @brief Type of property
     * @var string $type
     */
    public $type;

    /**
     * @brief Current value of the property
     * @var string $value
     */
    public $value;

    /**
     * @brief Minimum value of the property
     * @var string $min
     */
    public $min;

    /**
     * @brief Maximum value of the property
     * @var string $max
     */
    public $max;

    /**
     * @brief Step value of the property
     * @var string $step
     */
    public $step;

    /**
     * @brief Check if the property is good (optional)
     * @var string $assert
     */
    public $assert;

    /**
     * @brief Mapping to properties (optional)
     * @var string $propertymap
     */
    public $propertymap;

    /**
     * @brief Code to execute if the property change (optional)
     * @var string $onchange
     */
    public $onchange;

    /**
     * @brief Description of the param (optional)
     * @var string $desc
     */
    public $desc;

    /**
     * @brief Array of enums if param property have different enums (optional)
     * @var array|PropertyEnum $propertyenums
     */
    public $propertyenums;

    /**
     * @brief Array of property class specify the high level properties
     * @var array|Property $properties
     */
    public $properties;

    /**
     * @brief Reference to the associated parent block (don't set if parentProperty is set)
     * @var Block $parentBlock
     */
    public $parentBlock;

    /**
     * @brief Reference to the associated parent Property (don't set if parentBlock is set)
     * @var Property $parentProperty
     */
    public $parentProperty;

    /**
     * @brief constructor of Property
     * 
     * Initialise all the internal members and call parse_xml if $xml is set
     * @param SimpleXMLElement|null $xml if it's different of null, call the
     * xml parser to fill members
     */
    function __construct($xml = null)
    {
        $this->parentBlock = null;
        $this->parentProperty = null;
        $this->propertyenums = array();
        $this->properties = array();
        if ($xml)
            $this->parse_xml($xml);
    }

    /**
     * @brief funtion that export as string the main content of the class
     * instance
     * @return string
     */
    public function __toString()
    {
        $string = $this->name
                . " caption:'" . $this->caption
                . "' type:'" . $this->type
                . "' value:'" . $this->value
                . "' min:'" . $this->min
                . "' max:'" . $this->max
                . "' step:'" . $this->step
                . "' desc:'" . $this->desc
                . "'";
        return $string;
    }

    /**
     * @brief internal function to fill this instance from input xml structure
     * 
     * Can be call only from this node into the constructor
     * @param SimpleXMLElement $xml xml element to parse
     */
    protected function parse_xml($xml)
    {
        $this->name = (string) $xml['name'];
        if (strpos($this->name, '.') !== false)
            error("Property name cannot contains . (dot) in \"$this->name\"", 5, "Property");
        $this->caption = (string) $xml['caption'];
        $this->type = (string) $xml['type'];
        $this->value = (string) $xml['value'];
        $this->min = (string) $xml['min'];
        $this->max = (string) $xml['max'];
        $this->step = (string) $xml['step'];
        $this->assert = (string) $xml['assert'];
        $this->propertymap = (string) $xml['propertymap'];
        $this->onchange = (string) $xml['onchange'];
        $this->desc = (string) $xml['desc'];

        // enums
        if (isset($xml->enums))
        {
            foreach ($xml->enums->enum as $enumXml)
            {
                $this->addPropertyEnum(new PropertyEnum($enumXml));
            }
        }

        // properties
        if (isset($xml->properties))
        {
            foreach ($xml->properties->property as $propertyXml)
            {
                $this->addSubProperty(new Property($propertyXml));
            }
        }
    }

    /**
     * @brief permits to output this instance
     * 
     * Return a formated node for the node_generated file. This method call all
     * the children getXmlElement to add into this node.
     * @param DOMDocument $xml reference of the output xml document
     * @param string $format desired output file format
     * @return DOMElement xml element corresponding to this current instance
     */
    public function getXmlElement($xml, $format)
    {
        $xml_element = $xml->createElement("property");

        // name
        $att = $xml->createAttribute('name');
        $att->value = $this->name;
        $xml_element->appendChild($att);

        // value
        $att = $xml->createAttribute('value');
        $att->value = $this->value;
        $xml_element->appendChild($att);

        if ($format == "complete" or $format == "blockdef")
        {
            // caption
            $att = $xml->createAttribute('caption');
            $att->value = $this->caption;
            $xml_element->appendChild($att);

            // type
            $att = $xml->createAttribute('type');
            $att->value = $this->type;
            $xml_element->appendChild($att);

            // min
            $att = $xml->createAttribute('min');
            $att->value = $this->min;
            $xml_element->appendChild($att);

            // max
            $att = $xml->createAttribute('max');
            $att->value = $this->max;
            $xml_element->appendChild($att);

            // step
            $att = $xml->createAttribute('step');
            $att->value = $this->step;
            $xml_element->appendChild($att);

            // assert
            $att = $xml->createAttribute('assert');
            $att->value = $this->assert;
            $xml_element->appendChild($att);

            // propertymap
            $att = $xml->createAttribute('propertymap');
            $att->value = $this->propertymap;
            $xml_element->appendChild($att);

            // onchange
            $att = $xml->createAttribute('onchange');
            $att->value = $this->onchange;
            $xml_element->appendChild($att);

            // desc
            $att = $xml->createAttribute('desc');
            $att->value = $this->desc;
            $xml_element->appendChild($att);

            // propertyenums
            if (!empty($this->propertyenums))
            {
                $xml_propertyenums = $xml->createElement("enums");
                foreach ($this->propertyenums as $property)
                {
                    $xml_propertyenums->appendChild($property->getXmlElement($xml, $format));
                }
                $xml_element->appendChild($xml_propertyenums);
            }
        }

        // properties
        if (!empty($this->properties))
        {
            $xml_property = $xml->createElement("properties");
            foreach ($this->properties as $property)
            {
                $xml_property->appendChild($property->getXmlElement($xml, $format));
            }
            $xml_element->appendChild($xml_property);
        }

        return $xml_element;
    }

    /**
     * @brief Add a property enum to the block 
     * @param PropertyEnum $propertyenum property enum to add to the property
     */
    function addPropertyEnum($propertyenum)
    {
        $propertyenum->parentProperty = $this;
        array_push($this->propertyenums, $propertyenum);
    }

    /**
     * @brief return a reference to the property enum with the name $name, if not
     * found, return false
     * @param string $name name of the property enum to search
     * @param bool $casesens take care or not of the case of the name
     * @return PropertyEnum found property enum
     */
    function getPropertyEnum($name, $casesens = true)
    {
        if ($casesens)
        {
            foreach ($this->propertyenums as $propertyenum)
            {
                if ($propertyenum->name == $name)
                    return $propertyenum;
            }
        }
        else
        {
            foreach ($this->propertyenums as $propertyenum)
            {
                if (strcasecmp($propertyenum->name, $name) == 0)
                    return $propertyenum;
            }
        }
        return null;
    }

    /**
     * @brief delete a property enum from his name
     * @param string $name name of the property enum to delete 
     */
    function delPropertyEnum($name)
    {
        $i = 0;
        foreach ($this->propertyenums as $propertyenum)
        {
            if ($propertyenum->name == $name)
            {
                unset($this->propertyenums[$i]);
                return;
            }
            $i++;
        }
    }

    /**
     * @brief Add a sub-property to the property 
     * @param Property $property sub-property to add to the property
     */
    function addSubProperty($property)
    {
        $property->parentProperty = $this;
        array_push($this->properties, $property);
    }

    /**
     * @brief alias to addSubProperty($property)
     * @param Property $property sub-property to add to the property
     */
    function addProperty($property)
    {
        $this->addSubProperty($property);
    }

    /**
     * @brief return a reference to the property with the name $name, if not found, return false
     * @param string $name name of the property enum to search
     * @param bool $casesens take care or not of the case of the name
     * @return Property found property
     */
    function getSubProperty($name, $casesens = true)
    {
        if ($casesens)
        {
            foreach ($this->properties as $property)
            {
                if ($property->name == $name)
                    return $property;
            }
        }
        else
        {
            foreach ($this->properties as $property)
            {
                if (strcasecmp($property->name, $name) == 0)
                    return $property;
            }
        }
        return null;
    }

    /**
     * @brief alias to getSubProperty($name, $casesens)
     * @param string $name name of the property enum to search
     * @param bool $casesens take care or not of the case of the name
     * @return Property found property
     */
    function getProperty($name, $casesens = true)
    {
        return $this->getSubProperty($name, $casesens);
    }

    /**
     * @brief delete a property from his name
     * @param string $name name of the property to delete 
     */
    function delSubProperty($name)
    {
        $i = 0;
        foreach ($this->properties as $property)
        {
            if ($property->name == $name)
            {
                unset($this->properties[$i]);
                return;
            }
            $i++;
        }
    }

    /**
     * @brief alias to delSubProperty($name)
     * @param string $name name of the property to delete 
     */
    function delProperty($name)
    {
        return $this->delSubProperty($name);
    }

    /**
     * @brief return the path of the property (without the name of the block)
     * @return string path of the property
     */
    function path()
    {
        if ($this->parentProperty == NULL)
            return $this->name;
        else
            return $this->parentProperty->path() . "." . $this->name;
    }

    /**
     * @brief Returns the complete path for the property with the name of the
     * block.
     * 
     * If the property is a subproperty of another property, the call is
     * recursive.
     * @return string complete path of the property
     */
    function completePath()
    {
        if ($this->parentProperty == NULL)
        {
            if (get_class($this->parentBlock) == "Flow")
                return $this->parentBlock->parentBlock->name . "." . $this->parentBlock->name . "." . $this->name;
            else
                return $this->parentBlock->name . "." . $this->name;
        }
        else
            return $this->parentProperty->path() . "." . $this->name;
    }

    /**
     * @brief Returns all the depending property of the expression $expression
     * @param string $expression expression that need to be analysed
     * @return array|string array of property path
     */
    static function dependsProperties($expression)
    {
        preg_match_all("/([a-zA-Z_]+[a-zA-Z0-9_]*\\.?)+/x", $expression, $matches);
        $depends = array();
        foreach ($matches[0] as $match)
        {
            if (substr($match, -6) === ".value" or substr($match, -5) === ".bits")
                array_push($depends, $match);
        }
        return array_unique($depends);
    }

    /**
     * @brief Transforms an expression into a global expression in the context
     * of $blockName
     * 
     * This function use Property::dependsProperties function.
     * @param string $expression expression to convert
     * @param string $blockName context of expression (name of the block)
     * @return string processed expression in global context
     */
    static function localToGlobalPropertyPath($expression, $blockName)
    {
        foreach (Property::dependsProperties($expression) as $depend)
        {
            $expression = str_replace($depend, $blockName . '.' . $depend, $expression);
        }
        return $expression;
    }

    /**
     * @brief Transforms the propertymap expression into a global expression in
     * the context of $blockName
     * @param string $blockName context of expression (name of the block)
     */
    function toGlobalPropertyPath($blockName)
    {
        $this->propertymap = Property::localToGlobalPropertyPath($this->propertymap, $blockName);
        $this->onchange = Property::localToGlobalPropertyPath($this->onchange, $blockName);

        foreach ($this->properties as $property)
        {
            $property->toGlobalPropertyPath($blockName);
        }
    }
}
